### ISSUE TITLE
Number of tasks per submitted job

### ISSUE THREAD

**Issue by gaow** [vatlab/SOS#602]
Suppose I have this situation:

```
[1]
parameter: seed = [x for x in range(1000)]
input: for_each = ['seed']
task:
R:
  set.seed(${_seed})
  ...
```

This will generate 1000 tasks, therefore 1000 HPC jobs. This sounds like a burden in many ways. I wonder if we can have some task option:

```
task: task_per_job = 20
```
and, 

```
task: total_jobs = 50
```

In the first case, SoS will put 20 tasks in one HPC job and submit 50 jobs. In the 2nd case, we tell SoS that we want to submit 50 jobs total and SoS will figure out how many tasks per job, in this case is 20. 

I under stand we do have the `max_job` option. But that still will result in submitting 1000 jobs in the end, though sequentially reather than at once. I think these two options would be great features to have.

**Comment by BoPeng** [comment]
Your script is wrong `for_each='seed'` but I get your point. It is actually a very good idea because one of the problems with the clusters is that they are not friendly to small jobs. I think `task_per_job` makes more sense although another name like `batch` might be simpler.

**Comment by gaow** [comment]
Sorry for that typo. Well I can see values in both `task_per_job` and `total_jobs`. For 2, it provides a different, more controlled flavor of `max_running_jobs` that has clear pros (and possibly cons as well). 

I'm not sure about `batch`. It sounds more like `total_jobs`. `batch = 20`  says that I run the tasks in 20 batches, each batch has 50 tasks.

**Comment by BoPeng** [comment]
How about we define some `bucket` for task queues? Something like:

1. PBS defines bucket size (cores, mem, walltime etc).
2. When PBS tries to submit jobs, it tries to fit multiple jobs into buckets according to task properties.
3. Multiple jobs are executed either in parallel or sequentially depending on the shape of bucket.

A bucket is actually a special task that execute other tasks. It can be made transparent in the sense that users only see individual tasks and their properties.

**Comment by gaow** [comment]
It sounds interesting but I am having trouble wrapping my head around it. Are there any examples you can draft for buckets that deals with my example of submitting 1000 tasks in 20 batches if I want to be very specific about these numbers?

**Comment by BoPeng** [comment]
It resembles cargo container. Say,

```
tasks: walltime='00:10:00', mem='10G', cores=1
```
and the container has

```
container_mem: 32G
container_walltime: '10:00:00'
container_cores: 8
```
Then the task engine will put certain amount of tasks into one job, executing 3 tasks at a time (< 32G), and up to 180 tasks (done in 10 hours) in total. The PBS system will only see these big containers, not individual tasks, but users will see the status of each tasks.

**Comment by BoPeng** [comment]
Implementation wise we do not need a super clever algorithm. Something like this should work:

1. step submits tasks. After all tasks are submitted, send an all-done signal.
2. work engine accepts tasks and fill a bucket.
3. If bucket is full, or cannot fit another task, or if an all-done signal is received, submit the bucket as a single job. 

It is possible to define multiple buckets in different shapes, or do a global optimization for smallest number of buckets, but that does not sound necessary to me, at least not for now.

**Comment by BoPeng** [comment]
One of the advantages of pre-pack method is that there will be less tasks to monitor.

**Comment by BoPeng** [comment]
This request is non-trivial. In any case there need to be a "container task" for other tasks and the handling of such tasks can be troublesome because changes to all the task preparation, result gathering code, and individual tasks can fail. Proper implementation would need to abstract task definition to allow for container tasks, even nested tasks.

Interface wise, this can be implemented at either

* task engine: `bucket_mem`, `bucket_walltime`, and `bucket_cores`
* task options: `stack=n` for `n` sequential runs, or `stack=(n, m)` for `n*m` jobs in `m` concurrent processes. `total_jobs` does not seem necessary because it can be derived from total number of tasks.

I can see a later tag on this ticket.

**Comment by gaow** [comment]
Just got back from a seminar ... A fixed bucket sounds fine. 

What if whenever a bucket is used we treat it as a single task, as if we write multiple commands in one bash script? It should not mess up our signature management etc. If the whole task fails it is the user's duty to figure out which one is the cause. But when they fix it and run again there wouldnt be unnecessary reruns as we keep track of files, not tasks. Is that right?

**Comment by BoPeng** [comment]
It depends a bit on how you would like the subtasks to be executed. If you simply want to run them sequentially, we can create a master script to call the task scripts sequentially. If you would like to make use of multi-processing etc, a more sophisticated method would be needed, and a master bash script might or might not be enough.

**Comment by BoPeng** [comment]
And, remember that the `job_template` is only defined for a single task, so if we keep this deign, we will need the task engine to create a container task and merge the individual tasks before we send it through the `job_template`. Then, during the execution, the tasks need to have their own pulse threads so that individual tasks could be monitored as usual.

This task engine implementation also does not solve the problem of task overhead because managing and monitor tasks can be resource intensive and with 10000 tasks, a `task status` run can take a long time and the Jupyter notebook can be overwhelmed as well. That is why I cannot give up the task option idea because in that case we could get a much smaller number of tasks.

Finally, if both methods require a container task model, we might be able to implement both. It will be a lot of work so I tend to postpone it.

**Comment by gaow** [comment]
I understand it is not an optimal solution to run subtasks sequentially, but the total amount of run time might be similar; and it can immediately reduce the number of jobs I submit and I'm cool with that. In fact perhaps the idea of bucket is too complicated compared to the possibly subtle difference between it vs a master script implementation.

**Comment by BoPeng** [comment]
Then we can fall back to a single-task design. Basically, we merge the input, depends, output, and task of several tasks, add their cores and create a master task (without subtasks). Are you ok with the
```
stack = n
stack = (n,m)
```
interface?

**Comment by gaow** [comment]
I'm trying to understand `stack = (n, m)` ... how does that fit into my 1000 simulation, 20 per job and 50 jobs example? I see it can be done by `stack = n`. Not sure what part `m` plays here. Sorry I keep using that simple example but I just want to make sure we are intuitive for simple user cases :)

**Comment by BoPeng** [comment]
`m` concurrent tasks, so if you have 

```
input: for_each={'i': range(1000)}
task:    stack=(20, 10), mem='1G', walltime='00:00:60', cores=1
```

You are creating

* 5 master tasks
* each master task handle 200 tasks with 10 concurrent jobs.
* each master task will use `cores=10`, `mem=10G`, and `walltime='00:20:60'`.

**Comment by BoPeng** [comment]
The cores might have to be `11` is there is a master process and `10` workers. As you can imagine, it would implemented as a worker pool of `10` workers.

**Comment by gaow** [comment]
Ah I see now. It is smart, and useful for jobs without multi-thread built in it (am so much used to take that for granted)! I think this could be our preferred solution over buckets, because when we do resort to this option we might already worry about having too much tasks to monitor.

**Comment by BoPeng** [comment]
Because workers can consume different number of subtasks, perhaps the parameter could be written as

```
stack=(200, 10)
```
or 
```
stack_size=200
stack_workers = 10
```

**Comment by gaow** [comment]
I agree `stack = (200, 10)` is cleaner, because I can immediately tell that I'll see 5 jobs with my `qstat`. 

So with interface, if I want 50 jobs then I should do `stack = (20, *)` where `*` is whatever I want SoS to help me with another layer of parallazation using multiple cores.

**Comment by BoPeng** [comment]
Yes.

**Comment by BoPeng** [comment]
But then `stack=(200, 10)` is not what I imaged as a stack of bricks with height 200 and width 10, the parameter would better be called `group_by` or `group`.

**Comment by BoPeng** [comment]
Take off the later tag. I actually planned for a feature freeze but this is something that I myself really like.

**Comment by gaow** [comment]
Yes I saw the great new beta release! What do you mean by feature freeze -- until when? 

It's my bad that I've just really started using it for remote jobs. I did not have enough motivation earlier because the lab has a 28 core computer that is very lightly used so I get most of my stuff done there. But now I am at the stage of data analysis for a project that requires need more computing powers. 

May I try it for another week from now, just in case there are things unique to my slurm platform that may be relevant?

**Comment by BoPeng** [comment]
Feature freeze means a stage where no new feature is allowed. I am aiming at a usable version (much less bugs) after this month.

**Comment by BoPeng** [comment]
Some quick implementation, 

```

[10]

input: for_each={'i': range(100)}
task: group_by = (10, 5)
sh:
    echo ${i} >> ${i}.txt
```

The overhead is still big because of the use of worker pools, but is perhaps tolerable for large number of tasks.

**Comment by BoPeng** [comment]
Separate option `group_by` to `trunk_size` and `trunk_workers`.

**Comment by BoPeng** [comment]
Documented [here](http://vatlab.github.io/SOS/doc/documentation/External_task.html#Option-trunk_size-187).

**Comment by gaow** [comment]
Wonderful! I'll test them out later this afternoon and report any issues.

**Comment by BoPeng** [comment]
Note that there is an inevitable file format change so you cannot use the updated sos to check and execute old tasks.
